miniapps/autodiff/admfem.hpp

```diff
@@ -0,0 +1,587 @@
+    /// values in vector uu. The result is returned in vector rr.
+    void QVectorFunc(const mfem::Vector &vparam, mfem::Vector &uu, mfem::Vector& rr)
+    {
+        CTD<double,const mfem::Vector, mfem::Vector,
+                        vector_size, state_size, param_size> tf;
+        tf(vparam,uu,rr);
+    }
+
+    /// Returns the gradient of CTD(...) in the dense matrix jac.
+    /// The dimensions of jac are vector_size x state_size, where state_size is the
+    /// length of vector uu.
+    void QJacobian(mfem::Vector &vparam, mfem::Vector &uu, mfem::DenseMatrix &jac)
+    {
+#ifdef MFEM_USE_ADFORWARD
+        // use forward mode
+
+        CTD<ad::ADFloatType, const Vector, ad::ADVectorType,
+                        vector_size, state_size, param_size> tf;
+        jac.SetSize(vector_size, state_size);
+        jac = 0.0;
+        {
+            ad::ADVectorType aduu(state_size);
+            ad::ADVectorType rr(vector_size);
+            for(int i=0;i<state_size;i++){
+                aduu[i].setValue(uu[i]);
+                aduu[i].setGradient(0.0);
+            }
+
+            for(int ii=0;ii<state_size;ii++){
+                aduu[ii].setGradient(1.0);
+                tf(vparam,aduu,rr);
+                for(int jj=0;jj<vector_size;jj++)
+                {
+                    jac(jj,ii)=rr[jj].getGradient();
+                }
+                aduu[ii].setGradient(0.0);
+            }
+        }
+#else //end MFEM_USE_ADFORWARD
+        // use reverse mode
+        CTD<ad::ADFloatType, const Vector, ad::ADVectorType,
+                vector_size, state_size, param_size> tf;
+
+        jac.SetSize(vector_size, state_size);
+        jac = 0.0;
+        {
+            ad::ADVectorType aduu(state_size);
+            ad::ADVectorType rr(vector_size);
+            for(int i=0;i<state_size;i++){
+                aduu[i]=uu[i];
+            }
+
+            ad::ADFloatType::TapeType& tape =ad::ADFloatType::getGlobalTape();
+            typename ad::ADFloatType::TapeType::Position pos=tape.getPosition();
+
+            tape.setActive();
+            for(int ii=0;ii<state_size;ii++){ tape.registerInput(aduu[ii]); }
+            tf(vparam,aduu,rr);
+            for(int ii=0;ii<vector_size;ii++){ tape.registerOutput(rr[ii]); }
+            tape.setPassive();
+            for(int jj=0;jj<vector_size;jj++){
+                rr[jj].setGradient(1.0);
+                tape.evaluate();
+                for(int ii=0;ii<state_size;ii++){
+                    jac(jj,ii)=aduu[ii].getGradient();
+                }
+                tape.clearAdjoints();
+                rr[jj].setGradient(0.0);
+            }
+            tape.reset(pos);
+        }
+#endif
+    }
+};
+
+template<template<typename, typename, typename, int, int> class CTD
+         , int state_size=1, int param_size=0>
+class QFunctionAutoDiff
+{
+public:
+
+    /// Evaluates a function for arguments vparam and uu.
+    /// The evaluatin is based on the operator() in the
```

Comment:
evaluatin -> evaluation

Sometimes you use `mfem::Vector` in this file, and sometimes you use just `Vector`.

Reply:
modified

I prefer mfem::Vector in order to show where the definition comes from.